src/freenet/node/NewPacketFormat.java

```diff
@@ -1015,8 +1015,17 @@ public long timeNextUrgent(boolean canSend) {
 		// Check for acks.
 		ret = Math.min(ret, timeCheckForAcks());
 		
-		// Always wake up after half an RTT, check whether stuff is lost or needs ack'ing.
-		ret = Math.min(ret, System.currentTimeMillis() + Math.min(100, (long)averageRTT()/2));
+		if(ret > now) {
+		    // Always wake up after half an RTT, check whether stuff is lost or needs ack'ing.
+		    ret = Math.min(ret, now + Math.min(100, (long)averageRTT()/2));
+		    
+		    if(canSend && DO_KEEPALIVES) {
```

Comment:
The if() means if we're going to wake up immediately we don't need to consider keepalives. If there is room to send in the message window, we check for whether we need to send a keepalive, i.e. using the time since we last sent a packet.